src/Microsoft.AspNetCore.Server.IISIntegration/AuthenticationHandler.cs

```diff
@@ -51,13 +52,26 @@ private WindowsPrincipal GetUser()
                     NativeMethods.CloseHandle(handle);
 
                     _context.Response.RegisterForDispose(winIdentity);
+                    // We don't want loggers accessing a disposed identity.
+                    // https://github.com/aspnet/Logging/issues/543#issuecomment-321907828
+                    _context.Response.OnCompleted(ClearUserDelegate, _context);
                     _user = new WindowsPrincipal(winIdentity);
                 }
             }
 
             return _user;
         }
 
+        private static Task ClearUser(object arg)
+        {
+            var context = (HttpContext)arg;
+            if (context.User is WindowsPrincipal)
+            {
+                context.User = null;
```

Comment:
Don't we guarantee HttpContext.User to be non null?  Should we just set it to a new ClaimsPrincipal?  Or will the context do that automatically if someone asks for it?

Reply:
DefaultHttpContext handles that. https://github.com/aspnet/HttpAbstractions/blob/5128efe0e88eaa176d4231ff62d754de59bb0edd/src/Microsoft.AspNetCore.Http/DefaultHttpContext.cs#L131-L143